### DISCUSSION TITLE
Selecting the equivalent of an array slice in the put DSL

### DISCUSSION THREAD

**Discussion by trantor** [johnkerl/miller#640] [Q&A] [answered]
Hey there.

I am having a bit of a problem with the DSL here. Provided that the string manipulation functionality is often rather clunky in miller (haven't looked at version 6, hopefully it gets better :D ) I am stumped looking in the documentation for what I described in the title.
I meant to do something like
```
 mlr -N --csv put '$host = joinv(splitnvx($2,"-")[1,2],"-")'
```
to split string on `-` and reassemble it to get just the 2 initial pieces, but I am stumped as to how to achieve that, since `[1,2]` does not seem to be cut it. I get that `mlr` does not technically have arrays but rather maps of sorts, however it's pretty bad that I am unable to find a clear answer or example about this even after spending 10 minutes or more looking.

I think that the answer, after scrolling through all the functions, might be to use `mapselect` to filter the map, but it's not exactly user-friendly as an approach.

@johnkerl what am I missing? 

*update*
```
mlr -N --csv put '$host = joinv(mapselect(splitnvx($2,"-"),1,2),"-")'
```
does what I need, but I believe my concern on the clarity of such a construct and/or its documentation as a pattern might be valid anyway.

**Comment by johnkerl** [comment, marked as answer]
Thanks @trantor -- Miller pre-6's lack of arrays and slicing is indeed frustrating. I ended up doing lots of things to make integer-keyed maps fill some gaps but (as you found) it's awkward.

In Miller 5 I think your `$host = joinv(mapselect(splitnvx($2,"-"),1,2),"-")` is about as good as it gets.

In Miller 6 (using the [REPL](https://johnkerl.org/miller6/repl/) to interact):

```
$ mlr repl
Miller v6.0.0-dev REPL for darwin:amd64:go1.16.5
Pre-release docs for Miller 6: https://johnkerl.org/miller6
Type ':h' or ':help' for online help; ':q' or ':quit' to quit.

[mlr] host = "ab-cd-efg-hijk"

[mlr] splita(host, "-")
["ab", "cd", "efg", "hijk"]

[mlr] pieces = splita(host, "-")

[mlr] pieces[1:2]
["ab", "cd"]

[mlr]
```

So, in short, `joinv(splita("ab-cd-efg-hijk", "-")[1:2], "-")` is `"ab-cd"`.

As a side note, strings can be sliced in Miller 6, as well.

You can check out the Miller-6 arrays page: https://johnkerl.org/miller6/reference-main-arrays/

The MIller-6 strings page is still a bit under construction: https://johnkerl.org/miller6/reference-main-strings/

**Reply by trantor** [reply]
Thanks @johnkerl . Would making a specific example of the use of `mapselect` for this kind of purpose in the Miller 5 docs be worth the time at this point?

**Reply by johnkerl** [reply]
@trantor can't hurt ... :)

**Reply by johnkerl** [reply]
@trantor 

https://miller.readthedocs.io/en/latest/cookbook.html#splitting-a-string-and-taking-a-few-of-the-components

https://johnkerl.org/miller6/shapes-of-data/#splitting-a-string-and-taking-a-few-of-the-components

What do you think?
